Disruption rules in MockTransportService should match all bound addresses of a node.

ywelsch opened the pull request:
Currently, disruption rules apply only to the main bound address of a node. As consequence, integration tests that use Network Partitions must currently ensure that nodes are bound to a single local interface (e.g. 127.0.0.1 / IPv4) in order to correctly work. 

This PR changes the disruption rules to work on all transport addresses that are bound by a node.

Relates to #14625

bleskes commented on test-framework/src/main/java/org/elasticsearch/test/disruption/NetworkDelaysPartition.java:
```diff
@@ -78,10 +78,10 @@ public synchronized void startDisrupting() {
-                         DiscoveryNode node2, MockTransportService transportService2) {
-        transportService1.addUnresponsiveRule(node1, duration);
-        transportService1.addUnresponsiveRule(node2, duration);
+    void applyDisruption(TransportAddress address1, MockTransportService transportService1,
```
can we add java docs? this starts to be confusing...

bleskes commented on test-framework/src/main/java/org/elasticsearch/test/disruption/NetworkPartition.java:
```diff
@@ -179,11 +174,13 @@ public synchronized void startDisrupting() {
-                applyDisruption(discoveryNode1, transportService1, discoveryNode2, transportService2);
+                for (TransportAddress address1 : transportService1.boundAddress().boundAddresses()) {
+                    for (TransportAddress address2 : transportService2.boundAddress().boundAddresses()) {
+                        applyDisruption(address1, transportService1, address2, transportService2);
```
do we also want to have the publish address just to be sure? that's what nodes typically use to communicate with each other.

ywelsch replied:
publish address should normally be one of the bound transport addresses. If not, we run into some trouble in LookupTestTransport which uses publish address to determine transport. How about (instead of opening a can of worms) adding a check that requires publish address to be part of bound addresses, and fail otherwise?

bleskes replied:
I'm a bit reluctant because:

> publish address should normally be one of the bound transport addresses. 

is not always true (the reason why we make this distinction is to allow setting it to something else). How about throwing it all in a set and let the addresses de-dup themselves?

bleskes commented on test-framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java:
```diff
@@ -132,16 +151,34 @@ public void sendRequest(DiscoveryNode node, long requestId, String action, Trans
-    public void addFailToSendNoConnectRule(DiscoveryNode node, final String... blockedActions) {
-        addFailToSendNoConnectRule(node, new HashSet<>(Arrays.asList(blockedActions)));
+    public void addFailToSendNoConnectRule(TransportService transportService, final String... blockedActions) {
+        for (TransportAddress transportAddress : transportService.boundAddress().boundAddresses()) {
```
same here - what about publish address?

bleskes commented on test-framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java:
```diff
@@ -110,8 +119,18 @@ public Transport original() {
+     * Adds a rule that will cause every send request to fail, and each new connect since the rule
+     * is added to fail as well.
+     */
+    public void addFailToSendNoConnectRule(TransportAddress transportAddress) {
```
can we document transportAddress, how we extract addresses from it and how it will affect the sendRequest command (which uses a disco node)?

bleskes commented:
I think this is great. Good not to have to specifiy a host. I left some comments. Mostly around documentation. Also, it would be great to adda  note here (and to the commit) as to why we doing it - i.e., that unicast ping doesn't use disco nodes but rather constructs wrappers around explicit addresses.

ywelsch commented:
Pushed another set of changes.

- Removed TransportAddress from NetworkPartition.
- Added Javadoc to MockTransportService
- Added method `extractTransportAddresses(TransportService)` that combines publish and bound addresses.

bleskes commented on test-framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java:
```diff
@@ -132,16 +166,34 @@ public void sendRequest(DiscoveryNode node, long requestId, String action, Trans
-    public void addFailToSendNoConnectRule(DiscoveryNode node, final String... blockedActions) {
-        addFailToSendNoConnectRule(node, new HashSet<>(Arrays.asList(blockedActions)));
+    public void addFailToSendNoConnectRule(TransportService transportService, final String... blockedActions) {
+        for (TransportAddress transportAddress : extractTransportAddresses(transportService)) {
```
I think the extraction here is redundant no - just pass the transport service...

bleskes commented:
LGTM. Great!
